src/main/scala/firrtl/passes/Passes.scala

```diff
@@ -1128,24 +1128,66 @@ object ExpandWhens extends Pass {
    }
 }
 
+// This pass does various things to make FIRRTL graph legal
+object Legalize extends Pass {
+  def name = "Legalize"
+
+  def legalizeShiftRight (e: DoPrim): Expression = e.op match {
+    case SHIFT_RIGHT_OP => {
+      val amount = e.consts(0).toInt
+      val width = long_BANG(tpe(e.args(0)))
+      lazy val msb = width - 1
+      if (amount >= width) {
+        e.tpe match {
+          case t: UIntType => UIntValue(0, IntWidth(1)) // TODO make zero-width
+          case t: SIntType =>
+            DoPrim(BITS_SELECT_OP, e.args, Seq(msb, msb), SIntType(IntWidth(1)))
+          case t => error(s"Unsupported type ${t} for Primop Shift Right")
+        }
+      } else {
+        e
+      }
+    }
+    case _ => e
+  }
+  def run (c: Circuit): Circuit = {
+    def legalize_e (e: Expression): Expression = {
+      e map (legalize_e) match {
+        case e: DoPrim => legalizeShiftRight(e)
+        case e => e
+      }
+    }
+    def legalize_s (s: Stmt): Stmt = s map (legalize_s) map (legalize_e)
+    def legalize_m (m: Module): Module = m match {
+      case m: InModule => InModule(m.info, m.name, m.ports, legalize_s(m.body))
+      case m: ExModule => m
+    }
+    Circuit(c.info, c.modules.map(legalize_m), c.main)
+  }
+}
+
 object ConstProp extends Pass {
    def name = "Constant Propogation"
    var mname = ""
+
    def const_prop_e (e:Expression) : Expression = {
       e map (const_prop_e) match {
          case (e:DoPrim) => {
             e.op match {
                case SHIFT_RIGHT_OP => {
-                  (e.args(0)) match {
-                     case (x:UIntValue) => {
-                        val b = x.value >> e.consts(0).toInt
-                        UIntValue(b,tpe(e).as[UIntType].get.width)
+                  val amount = e.consts(0).toInt
+                  e.args(0) match {
+                     case x: UIntValue => {
+                        val v = x.value >> amount
+                        val w = (x.width - IntWidth(amount)) max IntWidth(1)
+                        UIntValue(v, w)
                      }
-                     case (x:SIntValue) => {
-                        val b = x.value >> e.consts(0).toInt
-                        SIntValue(b,tpe(e).as[SIntType].get.width)
+                     case x: SIntValue => { // take sign bit if shift amount is larger than arg width
```

Comment:
It turns out that >> in Scala does the right thing here so UInt and SInt can be treated the same and no special casing is necessary